seismicpro/src/weathering_velocity/weathering_velocity.py

```diff
@@ -0,0 +1,544 @@
+
+    @classmethod
+    def from_picking(cls, offsets, picking_times, n_layers=None, init=None, bounds=None, ascending_velocity=True,
+                 freeze_t0=False, **kwargs):
+        """Method fits the weathering model parameters from the offsets and the first break picking times.
+
+        Parameters
+        ----------
+        offsets : 1d ndarray
+            Offsets of the traces. Measured in meters.
+        picking_times : 1d ndarray
+            First break picking times. Measured in milliseconds.
+        init : dict, defaults to None
+            Initial parameters of a weathering model.
+        bounds : dict, defaults to None
+            Left and right bounds of the weathering model parameters.
+        n_layers : int, defaults to None
+            Number of layers of a weathering model.
+        ascending_velocity : bool, defaults to True
+            Keep the ascend of the fitted velocities from i-th layer to i+1 layer.
+        freeze_t0 : bool, defaults to False
+            Avoid the fitting `t0`.
+        kwargs : dict, optional
+            Additional keyword arguments to `scipy.optimize.minimize`.
+
+        Attributes
+        ----------
+        offsets : 1d ndarray
+            Offsets of traces. Measured in meters.
+        picking_times : 1d ndarray
+            Picking times of traces. Measured in milliseconds.
+        init : dict
+            The inital values used to fit the parameters of the weathering model. Includes the calculated non-passed
+            keys and values. Have the common key notation.
+        bounds : dict
+            The left and right bounds used to fit the parameters of the weathering model. Includes the calculated
+            non-passed keys and values. Have the common key notation.
+        n_layers : int
+            Number of the weathering model layers used to fit the parameters of the weathering model.
+        params : dict
+            The fitted parameters of a weathering model. Have the common key notation.
+
+        Raises
+        ------
+        ValueError
+            If any `init` values are negative.
+            If any `bounds` values are negative.
+            If left bound greater than right bound.
+            If init value is out of the bound interval.
+            If passed `init` and/or `bounds` keys are insufficient or excessive.
+            If an union of `init` and `bounds` keys less than 2 or `n_layers` less than 1.
+        """
+        self = cls()
+        init = {} if init is None else init
+        bounds = {} if bounds is None else bounds
+        self._check_values(init, bounds)
+
+        self.offsets = offsets
+        self.picking_times = picking_times
+        self.max_offset = offsets.max()
+
+        self.init = {**self._calc_init_by_layers(n_layers), **self._calc_init_by_bounds(bounds), **init}
+        self.bounds = {**self._calc_bounds_by_init(), **bounds}
+        self._check_keys(self.bounds)
+        self.n_layers = len(self.bounds) // 2
```

Comment:
I suggest to distinguish this parts, make only the necessary validation checks and do the calculations only if the check passed

```suggestion
init = {**self._calc_init_by_bounds(bounds), **init}

if n_layers is not None:
    self.validate_excessive(init)
    self.n_layers = n_layers
    init = {**self._calc_init_by_layers(n_layers), **init}
else:
    self.validate_insufficient(init)
    self.n_layers = len(init) / 2
self.bounds = {**self._calc_bounds_by_init(), **bounds}
```